### ISSUE TITLE
YoLo(v8) detector reports only one face per image ...

### ISSUE THREAD

**Issue by AndreaLanfranchi** [serengil/deepface#1130]
... while it should return all the faces detected on the image.

Consider [this code](https://github.com/serengil/deepface/blob/d7202cf70a6aa1ab4e9aa05a396f6a4f1559a4d8/deepface/detectors/Yolo.py#L72-L103) :
```python
        # Detect faces
        results = self.model.predict(img, verbose=False, show=False, conf=0.25)[0]

        # For each face, extract the bounding box, the landmarks and confidence
        for result in results:

            if result.boxes is None or result.keypoints is None:
                continue

            # Extract the bounding box and the confidence
            x, y, w, h = result.boxes.xywh.tolist()[0]
            confidence = result.boxes.conf.tolist()[0]

            # left_eye_conf = result.keypoints.conf[0][0]
            # right_eye_conf = result.keypoints.conf[0][1]
            left_eye = result.keypoints.xy[0][0].tolist()
            right_eye = result.keypoints.xy[0][1].tolist()

            # eyes are list of float, need to cast them tuple of int
            left_eye = tuple(int(i) for i in left_eye)
            right_eye = tuple(int(i) for i in right_eye)

            x, y, w, h = int(x - w / 2), int(y - h / 2), int(w), int(h)
            facial_area = FacialAreaRegion(
                x=x,
                y=y,
                w=w,
                h=h,
                left_eye=left_eye,
                right_eye=right_eye,
                confidence=confidence,
            )
            resp.append(facial_area)
```

This exposes a misunderstanding of the values returned by the `predict` method.
According to documentation the method returns a `List[Results]` where every element of the list is a complex type `Results` [which have a several members](https://docs.ultralytics.com/modes/predict/#working-with-results). 
In other words the model can receive as input multiple images (a 4 dimension numpy.ndarray) and return 1 Results (note the plural) for every processed image.

Now the problem in the above cited code is that :

1. `self.model.predict(...)[0]` picks the first element only of the outcome
2. the loop `for result in results` will be executed only once as, due to point 1, results have only one element
3. the consume of the values in `results` picks only the first element of each subcollection (note always index [0] is enforced)

As a result this loop can only return one detection for one picture but we cannot know which one as it depends on how the library parallelizes the detection task: it could be the main face or a minor face in the background.

**Comment by AndreaLanfranchi** [comment]
This has also an impact on how the pickle file from the dataset is generated as the processed images will only record the presence of 0 or 1 image: images with multiple detectable faces will have only one.

**Comment by serengil** [comment]
This is not true when I tested it

```python
from deepface import DeepFace

# https://raw.githubusercontent.com/serengil/deepface/master/tests/dataset/couple.jpg
img_path = "couple.jpg"

objs = DeepFace.extract_faces(img_path=img_path, detector_backend="yolov8")

assert len(objs) == 2
```

**Comment by AndreaLanfranchi** [comment]
Can you also visualize the results ?
I don't know how it can possibly work as the reading of the code is in deep contrast with ultrlytics documentation

**Comment by serengil** [comment]
![yolo-detection](https://github.com/serengil/deepface/assets/18491038/351696c3-ec58-4abb-ba8e-768b03d9dbe7)

**Comment by AndreaLanfranchi** [comment]
Odd ... to say the least

**Comment by AndreaLanfranchi** [comment]
I get that.
What does not compute (for me) is that results object is **one instance of an object** which has the following attributes
Results objects have the following attributes:

```
Attribute	Type	Description
orig_img	numpy.ndarray	The original image as a numpy array.
orig_shape	tuple	The original image shape in (height, width) format.
boxes	Boxes, optional	A Boxes object containing the detection bounding boxes.
masks	Masks, optional	A Masks object containing the detection masks.
probs	Probs, optional	A Probs object containing probabilities of each class for classification task.
keypoints	Keypoints, optional	A Keypoints object containing detected keypoints for each object.
obb	OBB, optional	An OBB object containing oriented bounding boxes.
speed	dict	A dictionary of preprocess, inference, and postprocess speeds in milliseconds per image.
names	dict	A dictionary of class names.
path	str	The path to the image file.
```

Hence the loop `for result in results` should iterate only 1 time

**Comment by serengil** [comment]
No, results variable has 2 items, that is why it will be iterated 2 times.

**Comment by AndreaLanfranchi** [comment]
Do you agree `self.model.predict(...)` returns a list ?
If we slice the list with `self.model.predict(...)[0]` then results has 1 item only

This is the signature of the predict function
```python
    def predict(
        self,
        source: Union[str, Path, int, list, tuple, np.ndarray, torch.Tensor] = None,
        stream: bool = False,
        predictor=None,
        **kwargs,
    ) -> list:
```

**Comment by serengil** [comment]
The following line is always returning a list with 1 items. Because as I explained, we are feeding img as (1, 224, 224, 3).

```
self.model.predict(...)
```

If the given image has 2 faces, then that 1 item has 2 items.

**Comment by serengil** [comment]
To make this clear: 

Suppose that you are feeding 2 images (img shape is (2, 224, 224, 3) in this case) to yolo and each image has 3 and 5 faces respectively.

self.model.predict(...) -> this will return a list with 2 items because you feed 2 images.

self.model.predict(...)[0] -> this will have 3 items because there are 3 faces in the first image.

self.model.predict(...)[1] -> this will have 5 items because there are 5 faces in the second image.

**Comment by serengil** [comment]
@AndreaLanfranchi do you have any concerns?

**Comment by AndreaLanfranchi** [comment]
Yes I have ... but still trying to understand where's the mess.
According to [documentation from ultralytics](https://docs.ultralytics.com/modes/predict/#working-with-results) :
> All Ultralytics predict() calls will return a list of Results objects
where each item of the list is a `Results` object (one object which holds subcollections) for each image processed.
Hence this `self.model.predict(...)[0]` is inequivocably a slice of a list (which may contain _N_ elements) to a single element.
As a result it is impossible to comprehend (for me) how possibly this
```en(self.model.predict(...)[0]) == 2``` 
can ever be true. 
Unless I miss some hidden and illogic semantic in python.

**Comment by serengil** [comment]
I explained this already actually.

```
Suppose that you are feeding 2 images (img shape is (2, 224, 224, 3) in this case) to yolo and each image has 3 and 5 faces respectively.

self.model.predict(...) -> this will return a list with 2 items because you feed 2 images.

self.model.predict(...)[0] -> this will have 3 items because there are 3 faces in the first image.

self.model.predict(...)[1] -> this will have 5 items because there are 5 faces in the second image.
```

**Comment by AndreaLanfranchi** [comment]
Nope ... you're interpreting (or probably observing empyrically) but is not what the documentation says.
Following your example:

self.model.predict(...) -> this **SHOULD** return a List[List[Results]] hence a total of 5 results
self.model.predict(...)[0] -> this should return a **slice** of List[List[Results]] where the outer list have only one item (the first) and the inner one has 3
self.model.predict(...)[0] -> this should return a **slice** of List[List[Results]] where the outer list have only one item (the second) and the inner one has 2

The discrepancy stands here

> All Ultralytics predict() calls will return a list of Results objects

hence List[Results] not List[List[Results]]

If I read the official doc (not your impl) the notation `List[Results][0]` is a list of 1 item only.
Then Results _can_ have multiple items accessible by properties `boxes`, `masks`, `probs`

Either the official doc is written with feet or there's something not correct.

**Comment by serengil** [comment]
```
Suppose that you are feeding 2 images (img shape is (2, 224, 224, 3) in this case) to yolo and each image has 3 and 5 faces respectively.
````

There are 8 faces in 2 images in my assumption.

You suppose `self.model.predict(...)` should return a list with 8 items. Then, how can you understand that how many faces available in the first image, and how many images available in second image?

Current implementation is returning a list with 2 items. You can find out how many faces available in each image.

**Comment by AndreaLanfranchi** [comment]
>You suppose self.model.predict(...) should return a list with 8 items. Then, how can you understand that how many faces available in the first image, and how many images available in second image?

Nope: this is not correct ... I'm assuming the return value is a List of lists i.e. a bidimensional array or, if you prefer, `List[List[type]]`
With this concept your results would have **first item of the results** to be a list with 3 items and **second item of the list** to be a list as well with 5 items.

Which is not actually the case. In fact, and here the documentation is obscure, the type `Results` embeds a _hidden_ list.
Looking now at it's internal implementation and I see :

```python
        [...]
        self._keys = "boxes", "masks", "probs", "keypoints", "obb"

    def __getitem__(self, idx):
        """Return a Results object for the specified index."""
        return self._apply("__getitem__", idx)

    def __len__(self):
        """Return the number of detections in the Results object."""
        for k in self._keys:
            v = getattr(self, k)
            if v is not None:
                return len(v)
```

So basically the `Results` type is a an implicit list which has a `len` property which value is the lenght of the first not None property (amongst the ones listed in keys). As a result the iterator behaves like this

```python
    def _apply(self, fn, *args, **kwargs):
        """
        Applies a function to all non-empty attributes and returns a new Results object with modified attributes. This
        function is internally called by methods like .to(), .cuda(), .cpu(), etc.

        Args:
            fn (str): The name of the function to apply.
            *args: Variable length argument list to pass to the function.
            **kwargs: Arbitrary keyword arguments to pass to the function.

        Returns:
            Results: A new Results object with attributes modified by the applied function.
        """
        r = self.new()
        for k in self._keys:
            v = getattr(self, k)
            if v is not None:
                setattr(r, k, getattr(v, fn)(*args, **kwargs))
        return r
```

Which, IMO extremely inefficiently, returns a new instance of `Results` on every loop.
I think I get it now.

As a result I'm keen to believe this code in the implementation 
```python
            # Extract the bounding box and the confidence
            x, y, w, h = result.boxes.xywh.tolist()[0]
            confidence = result.boxes.conf.tolist()[0]
```
carry a redundant (unnecessary) slicing of the first element `[0]`. In fact on every iteration only the pertaining box is returned.
You can close. Thank you.

**Comment by serengil** [comment]
I want to be more helpful but that is not my repo, sorry

**Comment by AndreaLanfranchi** [comment]
You can do this test on the couple.img

```python
        # Detect faces
        results = self.model.predict(img, verbose=False, show=False, conf=0.25) # Note I removed the slicing
        # print len(results) here ... it should give you 1
        # then slice 
        results = results[0]
        # now print len(results) after slicing ... it should give you 2
```
